Source/WebCore/Modules/mediastream/MediaDevices.cpp

```diff
@@ -121,6 +121,48 @@ bool MediaDevices::computeUserGesturePriviledge(GestureAllowedRequest requestTyp
     return isUserGesturePriviledged;
 }
 
+static std::optional<MediaConstraint> invalidRequiredConstraintForDeviceSelection(const MediaConstraints& constraints)
+{
+    // https://w3c.github.io/mediacapture-main/#dfn-allowed-required-constraints-for-device-selection
+    // The allowed required constraints for device selection contains the following constraint names:
+    // width, height, aspectRatio, frameRate, facingMode, resizeMode, sampleRate, sampleSize,
+    // echoCancellation, autoGainControl, noiseSuppression, latency, channelCount, deviceId, groupId.
+
+    std::optional<MediaConstraint> invalidConstraint;
+    constraints.mandatoryConstraints.filter([&invalidConstraint] (const MediaConstraint& constraint) mutable {
+        switch (constraint.constraintType()) {
+        case MediaConstraintType::Width:
+        case MediaConstraintType::Height:
+        case MediaConstraintType::AspectRatio:
+        case MediaConstraintType::FrameRate:
+        case MediaConstraintType::FacingMode:
+        case MediaConstraintType::SampleRate:
+        case MediaConstraintType::SampleSize:
+        case MediaConstraintType::EchoCancellation:
+        case MediaConstraintType::DeviceId:
+        case MediaConstraintType::GroupId:
+            break;
+
+        case MediaConstraintType::Volume:
+        case MediaConstraintType::FocusDistance:
+        case MediaConstraintType::Zoom:
+        case MediaConstraintType::WhiteBalanceMode:
+        case MediaConstraintType::Unknown:
+            invalidConstraint = { constraint };
+            break;
+
+        case MediaConstraintType::DisplaySurface:
+        case MediaConstraintType::LogicalSurface:
+            ASSERT_NOT_REACHED();
+            break;
+        }
+
+        return invalidConstraint.has_value();
```

Comment:
Is this check too harsh?
Maybe it would be simpler to do the checks in MediaTrackConstraintSet, by checking whether any of the extended constraints have exact/max/min members?